### ISSUE TITLE
Fail early (and tell user) on a message-link tap when the URL doesn't parse

### ISSUE THREAD

**Issue by chrisbobbe** [zulip/zulip-mobile#5518]
Like a toast saying "Link is invalid: ${url}".

We hope to do this just after #5492 settles.

**Comment by chrisbobbe** [comment]
Hmm, I think this might interfere with our planned fix for #4136.

I [expect](https://github.com/zulip/zulip-mobile/issues/4146#issuecomment-643001092) that a link that's gone through the URL parser will be URL encoded, and our current [plan](https://github.com/zulip/zulip-mobile/issues/4136#issuecomment-1294480513) for #4136 is to do something different from the standard URL encoding:

> So it seems like we need to, in `openLink` on iOS before passing the URL to `SafariView.show`:
> 
> * %-encode non-ASCII characters -- that's [Exception in SafariView on non-ASCII URLs ("unsupported scheme") #3315](https://github.com/zulip/zulip-mobile/issues/3315);
> * but _not_ %-encode `%` itself, instead leave it alone;
> * and presumably also leave alone all the other characters that `encodeURI` leaves alone, like `a` and `Z` and `/`;
> * and it's not clear if we should %-encode the remaining characters that `encodeURI` affects, like ` ` and `"` and a bunch of other punctuation.

This makes me wonder if there might be a better plan to consider, but I don't yet know what.

**Comment by gnprice** [comment]
> I [expect](https://github.com/zulip/zulip-mobile/issues/4146#issuecomment-643001092) that a link that's gone through the URL parser will be URL encoded, and our current [plan](https://github.com/zulip/zulip-mobile/issues/4136#issuecomment-1294480513) for #4136 is to do something different from the standard URL encoding:
> 
> > […]
>
> This makes me wonder if there might be a better plan to consider for #4136, but I don't yet know what.

Hmm, interesting.

Looking into a bit, I think it actually points us to such a plan -- more specifically, to a convenient and also reassuring way to implement that plan!

Specifically, if you take a URL string and round-trip it through the URL parser, you end up with pretty much the sort of partial percent-encoding I described -- non-ASCII characters get encoded, `%` doesn't:
```
> new URL('https://example.com/?a=b%3D&c=þ').href
'https://example.com/?a=b%3D&c=%C3%BE'
```
Contrast that with `encodeURI`:
```
> encodeURI('https://example.com/?a=b%3D&c=þ')
'https://example.com/?a=b%253D&c=%C3%BE'
```

As a note in the URL spec (https://url.spec.whatwg.org/#url-code-points) puts the behavior:
> Code points greater than U+007F DELETE will be converted to [percent-encoded bytes](https://url.spec.whatwg.org/#percent-encoded-byte) by the [URL parser](https://url.spec.whatwg.org/#concept-url-parser).

Concretely, what's happening in that example is that in [query state](https://url.spec.whatwg.org/#query-state), the input gets encoded using the [special-query percent-encode set](https://url.spec.whatwg.org/#special-query-percent-encode-set); and that contains all code points above U+007E, as do each of the other percent-encode sets defined in the standard, but does not contain U+0025 `%`, again like all the other percent-encode sets defined there.

I'll update #4136 with that thought.
